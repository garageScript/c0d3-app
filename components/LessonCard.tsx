import React from 'react'
import { useQuery } from '@apollo/react-hooks'
import { CheckCircle } from 'react-feather'
import { GET_SUBMISSIONS } from '../graphql/queries'
import {
  Props,
  ReviewButtonProps,
  ReviewCountProps
} from '../@types/lessonCard'
import NavLink from './NavLink'
import '../scss/lessonCard.scss'

const ReviewCount: React.FC<ReviewCountProps> = props => {
  if (props.shouldNotGetCount) {
    return null
  }

  const { loading, data } = useQuery(GET_SUBMISSIONS, {
    variables: {
      in: {
        id: `${props.lessonId}`
      }
    }
  })

  if (loading) {
    return (
      <div className="spinner-border spinner-border-sm mx-1" role="status" />
    )
  }

  if (!data) {
    return null
  }
  const pendingSubmissionsCount = data.submissions.reduce(
    (acc: number, val: any) => {
      if (val.status === 'open') {
        acc = acc + 1
        return acc
      }
      return acc
    },
    0
  )

  return <span> {pendingSubmissionsCount} </span>
}

const ReviewButton: React.FC<ReviewButtonProps> = props => {
  if (!props.isCompleted) {
    return null
  }
  return (
    <a
      href={props.reviewUrl}
      className="btn btn-sm bg-primary text-white float-right mb-2 mr-2"
    >
      Review
      <ReviewCount
        shouldNotGetCount={props.shouldNotGetCount}
        lessonId={props.lessonId}
      />
      Submissions
    </a>
  )
}

const LessonCard: React.FC<Props> = props => {
  const containerClass =
    props.currentState === 'inProgress'
      ? 'lesson-card__container_inprogress border-primary'
      : 'border-0'

  return (
    <div className={`card shadow-sm mt-3 ${containerClass}`}>
      <div className="d-flex p-2">
        <img
          src={`/curriculumAssets/lessonCoversSvg/${props.coverImg}`}
          alt={props.coverImg}
        />

        <div className="w-100 pl-4">
          {props.currentState === 'completed' && (
            <span className="badge badge-pill badge-success float-right mt-2 mr-2 p-2 d-flex align-items-center">
              <CheckCircle style={{ height: '15px' }} />
              <span className="mx-1">COMPLETED</span>
            </span>
          )}
          <div>
            <h4 className="lesson-card__title font-weight-bold mt-3">
              <NavLink
                as={`/curriculum/${props.lessonId}`}
                path="/curriculum/[lesson]"
                text={props.title}
              />
            </h4>
            <div>
              <div className="d-inline-block mr-4">
                <img
                  className="mr-2"
                  src="/curriculumAssets/icons/icon-challenge.svg"
                  alt="icon-challenge"
                />
                <span className="lesson-card__icon-text">
                  {props.challengeCount} CHALLENGES
                </span>
              </div>
            </div>
            <p className="lesson-card__description mt-2">{props.description}</p>
          </div>
          <ReviewButton
            isCompleted={props.currentState === 'completed'}
            reviewUrl={props.reviewUrl}
            shouldNotGetCount={props.shouldNotGetCount}
            lessonId={props.lessonId}
          />
        </div>
      </div>

      {props.currentState === 'inProgress' && (
        <div className="p-2 bg-primary">
          <NavLink
            path={props.docUrl}
            text="Start Lesson"
            className="lesson-card__button btn btn-light mr-2 my-1 text-primary"
            external
          />
          <NavLink
            path={props.challengesUrl}
            text="View Challenges"
            className="lesson-card__button btn bg-primary my-1 text-white border border-white"
          />
        </div>
      )}
    </div>
  )
}

export default LessonCard
