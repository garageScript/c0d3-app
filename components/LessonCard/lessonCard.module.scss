@use '../../scss/variables.module';

.lesson-card__container_inprogress {
  border-width: 2px;
}
.lesson-card__title {
  font-size: 1.5rem;
  line-height: 1.3;
  margin-bottom: 0.75rem;
}

.lesson-card__badge {
  align-self: flex-start;
  padding-left: 1em;
  padding-right: 1em;
  margin-left: auto;
  background-color: variables.$success;
  @media (max-width: 768px) {
    background-color: transparent;
    color: #00cf92;
    margin-left: auto;
    & > svg {
      width: 30px;
      height: 30px;
      stroke-width: 3px;
    }
  }
}

.lesson-card__icon-text {
  font-size: 0.8rem;
  line-height: 1.4;
}
.lesson-card__button {
  font-size: 13px;
  min-width: 134px;
}
.lesson-card__image {
  object-fit: contain;
}
.lesson-card__description {
  flex: 1;
}
