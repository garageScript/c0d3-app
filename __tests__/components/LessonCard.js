jest.mock('@apollo/react-hooks')
import * as React from 'react'
import LessonCard from '../../components/LessonCard'
import { useQuery } from '@apollo/react-hooks'
import { render } from '@testing-library/react'

describe('Lesson Card Complete State', () => {
  test('Should render lessonCard with null if no data', async () => {
    useQuery.mockReturnValue({
      data: null
    })

    const { container } = render(<LessonCard currentState="completed" />)
    expect(container).toMatchSnapshot()
  })
  test('Should render lessonCard with loading...', async () => {
    useQuery.mockReturnValue({
      loading: true
    })

    const { container } = render(<LessonCard currentState="completed" />)
    expect(container).toMatchSnapshot()
  })
  test('Should render lessonCard with submission count', async () => {
    useQuery.mockReturnValue({
      data: {
        submissions: [
          {
            status: 'open'
          },
          {
            status: 'open'
          },
          {
            status: 'needMoreWork'
          }
        ]
      }
    })

    const { container } = render(<LessonCard currentState="completed" />)
    expect(container).toMatchSnapshot()
  })
})

describe('Lesson Card', () => {
  test('Should render loading card when loading', async () => {
    useQuery.mockReturnValue({
      loading: true
    })

    const { container } = render(<LessonCard />)
    expect(container).toMatchSnapshot()
  })
  test('Should render ellpsis when no data is present', async () => {
    useQuery.mockReturnValue({
      loading: false
    })

    const { container } = render(<LessonCard />)
    expect(container).toMatchSnapshot()
  })
  test('Should render lessonCard with no submission count', async () => {
    useQuery.mockReturnValue({
      data: {
        submissions: [
          {
            status: 'underReview'
          },
          {
            status: 'underReview'
          }
        ]
      }
    })

    const { container } = render(<LessonCard />)
    expect(container).toMatchSnapshot()
  })
  test('Should render lessonCard with inProgress currentState prop', async () => {
    const { container } = render(<LessonCard currentState="inProgress" />)
    expect(container).toMatchSnapshot()
  })
  test('Should render lessonCard with lessonId prop', async () => {
    const { container } = render(<LessonCard lessonId={4} />)
    expect(container).toMatchSnapshot()
  })
})
