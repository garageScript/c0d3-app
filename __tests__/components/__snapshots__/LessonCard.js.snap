// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lesson Card Complete State Should render lessonCard with loading... 1`] = `
<div>
  <div
    class="card shadow-sm mt-3 d-block border-0"
  >
    <div
      class="d-flex p-2"
    >
      <img
        src="/curriculumAssets/lessonCoversSvg/undefined"
      />
      <div
        class="w-100 pl-4"
      >
        <span
          class="badge badge-pill badge-success float-right mt-2 mr-2 p-2 d-flex align-items-center"
        >
          <svg
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            style="height: 15px;"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M22 11.08V12a10 10 0 1 1-5.93-9.14"
            />
            <polyline
              points="22 4 12 14.01 9 11.01"
            />
          </svg>
          <span
            class="mx-1"
          >
            COMPLETED
          </span>
        </span>
        <div>
          <h4
            class="lesson-card__title font-weight-bold mt-3"
          >
            <a
              class=""
              href="/curriculum/undefined"
            />
          </h4>
          <div>
            <div
              class="d-inline-block mr-4"
            >
              <img
                alt="icon-challenge"
                class="mr-2"
                src="/curriculumAssets/icons/icon-challenge.svg"
              />
              <span
                class="lesson-card__icon-text"
              >
                 CHALLENGES
              </span>
            </div>
          </div>
          <p
            class="lesson-card__description mt-2"
          />
        </div>
        <a
          class="btn btn-sm bg-primary text-white float-right mb-2 mr-2"
        >
          Review
          <div
            class="spinner-border spinner-border-sm mx-1"
            role="status"
          />
          Submissions
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`Lesson Card Complete State Should render lessonCard with null if no data 1`] = `
<div>
  <div
    class="card shadow-sm mt-3 d-block border-0"
  >
    <div
      class="d-flex p-2"
    >
      <img
        src="/curriculumAssets/lessonCoversSvg/undefined"
      />
      <div
        class="w-100 pl-4"
      >
        <span
          class="badge badge-pill badge-success float-right mt-2 mr-2 p-2 d-flex align-items-center"
        >
          <svg
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            style="height: 15px;"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M22 11.08V12a10 10 0 1 1-5.93-9.14"
            />
            <polyline
              points="22 4 12 14.01 9 11.01"
            />
          </svg>
          <span
            class="mx-1"
          >
            COMPLETED
          </span>
        </span>
        <div>
          <h4
            class="lesson-card__title font-weight-bold mt-3"
          >
            <a
              class=""
              href="/curriculum/undefined"
            />
          </h4>
          <div>
            <div
              class="d-inline-block mr-4"
            >
              <img
                alt="icon-challenge"
                class="mr-2"
                src="/curriculumAssets/icons/icon-challenge.svg"
              />
              <span
                class="lesson-card__icon-text"
              >
                 CHALLENGES
              </span>
            </div>
          </div>
          <p
            class="lesson-card__description mt-2"
          />
        </div>
        <a
          class="btn btn-sm bg-primary text-white float-right mb-2 mr-2"
        >
          Review
          Submissions
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`Lesson Card Complete State Should render lessonCard with submission count 1`] = `
<div>
  <div
    class="card shadow-sm mt-3 d-block border-0"
  >
    <div
      class="d-flex p-2"
    >
      <img
        src="/curriculumAssets/lessonCoversSvg/undefined"
      />
      <div
        class="w-100 pl-4"
      >
        <span
          class="badge badge-pill badge-success float-right mt-2 mr-2 p-2 d-flex align-items-center"
        >
          <svg
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            style="height: 15px;"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M22 11.08V12a10 10 0 1 1-5.93-9.14"
            />
            <polyline
              points="22 4 12 14.01 9 11.01"
            />
          </svg>
          <span
            class="mx-1"
          >
            COMPLETED
          </span>
        </span>
        <div>
          <h4
            class="lesson-card__title font-weight-bold mt-3"
          >
            <a
              class=""
              href="/curriculum/undefined"
            />
          </h4>
          <div>
            <div
              class="d-inline-block mr-4"
            >
              <img
                alt="icon-challenge"
                class="mr-2"
                src="/curriculumAssets/icons/icon-challenge.svg"
              />
              <span
                class="lesson-card__icon-text"
              >
                 CHALLENGES
              </span>
            </div>
          </div>
          <p
            class="lesson-card__description mt-2"
          />
        </div>
        <a
          class="btn btn-sm bg-primary text-white float-right mb-2 mr-2"
        >
          Review
          <span>
             
            2
             
          </span>
          Submissions
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`Lesson Card Should render ellpsis when no data is present 1`] = `
<div>
  <div
    class="card shadow-sm mt-3 d-block border-0"
  >
    <div
      class="d-flex p-2"
    >
      <img
        src="/curriculumAssets/lessonCoversSvg/undefined"
      />
      <div
        class="w-100 pl-4"
      >
        <div>
          <h4
            class="lesson-card__title font-weight-bold mt-3"
          >
            <a
              class=""
              href="/curriculum/undefined"
            />
          </h4>
          <div>
            <div
              class="d-inline-block mr-4"
            >
              <img
                alt="icon-challenge"
                class="mr-2"
                src="/curriculumAssets/icons/icon-challenge.svg"
              />
              <span
                class="lesson-card__icon-text"
              >
                 CHALLENGES
              </span>
            </div>
          </div>
          <p
            class="lesson-card__description mt-2"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Lesson Card Should render lessonCard with inProgress currentState prop 1`] = `
<div>
  <div
    class="card shadow-sm mt-3 d-block lesson-card__container_inprogress border-primary"
  >
    <div
      class="d-flex p-2"
    >
      <img
        src="/curriculumAssets/lessonCoversSvg/undefined"
      />
      <div
        class="w-100 pl-4"
      >
        <div>
          <h4
            class="lesson-card__title font-weight-bold mt-3"
          >
            <a
              class=""
              href="/curriculum/undefined"
            />
          </h4>
          <div>
            <div
              class="d-inline-block mr-4"
            >
              <img
                alt="icon-challenge"
                class="mr-2"
                src="/curriculumAssets/icons/icon-challenge.svg"
              />
              <span
                class="lesson-card__icon-text"
              >
                 CHALLENGES
              </span>
            </div>
          </div>
          <p
            class="lesson-card__description mt-2"
          />
        </div>
      </div>
    </div>
    <div
      class="p-2 bg-primary"
    />
  </div>
</div>
`;

exports[`Lesson Card Should render lessonCard with lessonId prop 1`] = `
<div>
  <div
    class="card shadow-sm mt-3 d-block border-0"
  >
    <div
      class="d-flex p-2"
    >
      <img
        src="/curriculumAssets/lessonCoversSvg/undefined"
      />
      <div
        class="w-100 pl-4"
      >
        <div>
          <h4
            class="lesson-card__title font-weight-bold mt-3"
          >
            <a
              class=""
              href="/curriculum/4"
            />
          </h4>
          <div>
            <div
              class="d-inline-block mr-4"
            >
              <img
                alt="icon-challenge"
                class="mr-2"
                src="/curriculumAssets/icons/icon-challenge.svg"
              />
              <span
                class="lesson-card__icon-text"
              >
                 CHALLENGES
              </span>
            </div>
          </div>
          <p
            class="lesson-card__description mt-2"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Lesson Card Should render lessonCard with no submission count 1`] = `
<div>
  <div
    class="card shadow-sm mt-3 d-block border-0"
  >
    <div
      class="d-flex p-2"
    >
      <img
        src="/curriculumAssets/lessonCoversSvg/undefined"
      />
      <div
        class="w-100 pl-4"
      >
        <div>
          <h4
            class="lesson-card__title font-weight-bold mt-3"
          >
            <a
              class=""
              href="/curriculum/undefined"
            />
          </h4>
          <div>
            <div
              class="d-inline-block mr-4"
            >
              <img
                alt="icon-challenge"
                class="mr-2"
                src="/curriculumAssets/icons/icon-challenge.svg"
              />
              <span
                class="lesson-card__icon-text"
              >
                 CHALLENGES
              </span>
            </div>
          </div>
          <p
            class="lesson-card__description mt-2"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Lesson Card Should render loading card when loading 1`] = `
<div>
  <div
    class="card shadow-sm mt-3 d-block border-0"
  >
    <div
      class="d-flex p-2"
    >
      <img
        src="/curriculumAssets/lessonCoversSvg/undefined"
      />
      <div
        class="w-100 pl-4"
      >
        <div>
          <h4
            class="lesson-card__title font-weight-bold mt-3"
          >
            <a
              class=""
              href="/curriculum/undefined"
            />
          </h4>
          <div>
            <div
              class="d-inline-block mr-4"
            >
              <img
                alt="icon-challenge"
                class="mr-2"
                src="/curriculumAssets/icons/icon-challenge.svg"
              />
              <span
                class="lesson-card__icon-text"
              >
                 CHALLENGES
              </span>
            </div>
          </div>
          <p
            class="lesson-card__description mt-2"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
